client/fingerprint/env_gce.go

```diff
@@ -0,0 +1,231 @@
+	reqUrl := f.metadataURL + attribute
+	if recursive {
+		reqUrl = reqUrl + "?recursive=true"
+	}
+
+	parsedUrl, err := url.Parse(reqUrl)
+	if err != nil {
+		return "", err
+	}
+
+	req := &http.Request{
+		Method: "GET",
+		URL:    parsedUrl,
+		Header: http.Header{
+			"Metadata-Flavor": []string{"Google"},
+		},
+	}
+
+	res, err := f.client.Do(req)
+	if err != nil {
+		return "", err
+	}
+
+	resp, err := ioutil.ReadAll(res.Body)
+	res.Body.Close()
+	if err != nil {
+		f.logger.Printf("[ERR]: fingerprint.env_gce: Error reading response body for GCE %s", attribute)
+		return "", err
+	}
+
+	if res.StatusCode >= 400 {
+		return "", ReqError{res.StatusCode}
+	}
+
+	return string(resp), nil
+}
+
+func checkError(err error, logger *log.Logger, desc string) error {
+	// If it's a URL error, assume we're not actually in an GCE environment.
+	// To the outer layers, this isn't an error so return nil.
+	if _, ok := err.(*url.Error); ok {
+		logger.Printf("[ERR] fingerprint.env_gce: Error querying GCE " + desc + ", skipping")
+		return nil
+	}
+	// Otherwise pass the error through.
+	return err
+}
+
+func (f *EnvGCEFingerprint) Fingerprint(cfg *config.Config, node *structs.Node) (bool, error) {
+	if !f.isGCE() {
+		return false, nil
+	}
+
+	// newNetwork is populated and addded to the Nodes resources
+	newNetwork := &structs.NetworkResource{
+		Device: "eth0",
+	}
+
+	if node.Links == nil {
+		node.Links = make(map[string]string)
+	}
+
+	keys := []string{
+		"hostname",
+		"id",
+	}
+	for _, k := range keys {
+		value, err := f.Get(k, false)
+		if err != nil {
+			return false, checkError(err, f.logger, k)
+		}
+
+		// assume we want blank entries
+		node.Attributes["platform.gce."+k] = strings.Trim(string(value), "\n")
+	}
+
+	// These keys need everything before the final slash removed to be usable.
+	keys = []string{
+		"machine-type",
+		"zone",
+	}
+	for _, k := range keys {
+		value, err := f.Get(k, false)
+		if err != nil {
+			return false, checkError(err, f.logger, k)
+		}
+
+		index := strings.LastIndex(value, "/")
+		value = value[index+1:]
+		node.Attributes["platform.gce."+k] = strings.Trim(string(value), "\n")
+	}
+
+	// Get internal and external IP (if it exits)
+	value, err := f.Get("network-interfaces/0/ip", false)
+	if err != nil {
+		return false, checkError(err, f.logger, "ip")
+	}
+	newNetwork.IP = strings.Trim(value, "\n")
+	newNetwork.CIDR = newNetwork.IP + "/32"
+	node.Attributes["network.ip-address"] = newNetwork.IP
+
+	value, err = f.Get("network-interfaces/0/access-configs/0/external-ip", false)
+	if re, ok := err.(ReqError); err != nil && (!ok || re.StatusCode != 404) {
+		return false, checkError(err, f.logger, "external IP")
+	}
+	value = strings.Trim(value, "\n")
+	if len(value) > 0 {
+		node.Attributes["platform.gce.external-ip"] = value
+	}
+
+	var tagList []string
+	value, err = f.Get("tags", false)
+	if err != nil {
+		return false, checkError(err, f.logger, "tags")
+	}
+	err = json.Unmarshal([]byte(value), &tagList)
```

Comment:
if err := json.Unmarshal([]byte(value), &tagList); err != nil {
                f.logger.Printf("[WARN] fingerprint.env_gce: Error decoding instance tags: %s", err.Error())
} else {
               for _, tag := range tagList {
			node.Attributes["platform.gce.tag."+tag] = "true"
		}
}